src/main/java/makaroshyna/onlinebookstore/service/shoppingcart/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,78 @@
+package makaroshyna.onlinebookstore.service.shoppingcart;
+
+import jakarta.transaction.Transactional;
+import java.util.HashSet;
+import lombok.RequiredArgsConstructor;
+import makaroshyna.onlinebookstore.dto.cartitem.CartItemResponseDto;
+import makaroshyna.onlinebookstore.dto.cartitem.CreateCartItemRequestDto;
+import makaroshyna.onlinebookstore.dto.cartitem.UpdateCartItemRequestDto;
+import makaroshyna.onlinebookstore.dto.shoppingcart.ShoppingCartResponseDto;
+import makaroshyna.onlinebookstore.exception.EntityNotFoundException;
+import makaroshyna.onlinebookstore.mapper.CartItemMapper;
+import makaroshyna.onlinebookstore.mapper.ShoppingCartMapper;
+import makaroshyna.onlinebookstore.model.CartItem;
+import makaroshyna.onlinebookstore.model.ShoppingCart;
+import makaroshyna.onlinebookstore.model.User;
+import makaroshyna.onlinebookstore.repository.shoppingcart.ShoppingCartRepository;
+import makaroshyna.onlinebookstore.service.cartitem.CartItemService;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final CartItemMapper cartItemMapper;
+    private final CartItemService cartItemService;
+
+    @Override
+    public ShoppingCartResponseDto getByUserId(Long userId) {
+        return shoppingCartMapper.toDto(getShoppingCart(userId));
+    }
+
+    @Override
+    @Transactional
+    public CartItemResponseDto addToCart(CreateCartItemRequestDto requestDto, Long userId) {
+        ShoppingCart shoppingCart = getShoppingCart(userId);
+        CartItem cartItem = cartItemService.save(requestDto, shoppingCart);
+        shoppingCart.getCartItems().add(cartItem);
+        shoppingCartRepository.save(shoppingCart);
```

Comment:
I do not like the double save operation, a single save should be enough, just validate the book and that the shopping cart exists, set that shopping cart to the item and save the item, you can also use item repository directly here if you want, that would be fine